01 — Data Inspection

**Purpose:** Load the Telco Customer Churn dataset and understand
its structure before any cleaning or modeling.

**Dataset:** data/raw/telco_customer_churn.csv

In [4]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 200)

print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)

Pandas version: 2.3.3
NumPy version: 2.2.6


In [5]:
DATA_PATH = "../data/raw/telco_customer_churn.csv"

df = pd.read_csv(DATA_PATH)

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 7043
Columns: 21


In [6]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [8]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


In [9]:
missing = df.isnull().sum()
missing = missing[missing > 0]

if len(missing) == 0:
    print("No missing values detected by pandas.")
else:
    print("Missing values per column:")
    print(missing)

No missing values detected by pandas.


In [10]:
blank_total_charges = (df["TotalCharges"].astype(str).str.strip() == "").sum()

print("Blank TotalCharges entries:", blank_total_charges)
print("TotalCharges dtype:", df["TotalCharges"].dtype)

Blank TotalCharges entries: 11
TotalCharges dtype: object


In [11]:
churn_counts = df["Churn"].value_counts()
churn_pct = df["Churn"].value_counts(normalize=True) * 100

print("Churn counts:")
print(churn_counts)

print()
print("Churn percentage:")
print(churn_pct.round(2))

Churn counts:
Churn
No     5174
Yes    1869
Name: count, dtype: int64

Churn percentage:
Churn
No     73.46
Yes    26.54
Name: proportion, dtype: float64


In [12]:
categorical_cols = df.select_dtypes(include=["object"]).columns

print("Categorical columns:", len(categorical_cols))
print()

for col in categorical_cols:
    unique_vals = df[col].nunique()
    print(f"{col:25s} → {unique_vals} unique values")

Categorical columns: 18

customerID                → 7043 unique values
gender                    → 2 unique values
Partner                   → 2 unique values
Dependents                → 2 unique values
PhoneService              → 2 unique values
MultipleLines             → 3 unique values
InternetService           → 3 unique values
OnlineSecurity            → 3 unique values
OnlineBackup              → 3 unique values
DeviceProtection          → 3 unique values
TechSupport               → 3 unique values
StreamingTV               → 3 unique values
StreamingMovies           → 3 unique values
Contract                  → 3 unique values
PaperlessBilling          → 2 unique values
PaymentMethod             → 4 unique values
TotalCharges              → 6531 unique values
Churn                     → 2 unique values


In [13]:
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns

print("Numeric columns:", list(numeric_cols))
print()

for col in numeric_cols:
    print(
        f"{col:20s} "
        f"min={df[col].min()}   "
        f"max={df[col].max()}   "
        f"mean={df[col].mean():.2f}"
    )

Numeric columns: ['SeniorCitizen', 'tenure', 'MonthlyCharges']

SeniorCitizen        min=0   max=1   mean=0.16
tenure               min=0   max=72   mean=32.37
MonthlyCharges       min=18.25   max=118.75   mean=64.76


## Day 2 — Findings Summary

**Dataset shape:** 7,043 rows × 21 columns.

**Target column:** `Churn` (Yes / No), currently stored as text.

**Class distribution:** Imbalanced — approximately 26–27% of
customers churned. This must be considered when choosing
evaluation metrics later.

**Data quality issues found so far:**

1. `TotalCharges` is stored as text (`object`) instead of a number.
2. `TotalCharges` contains a small number of blank entries that
   pandas does not detect as missing values.
3. `customerID` is a unique identifier with no predictive value
   and should be removed before modeling.
4. `Churn` needs to be converted from Yes/No text to 1/0 for
   machine learning.

**Next step (Day 3):** Clean the dataset — fix data types, handle
blanks, drop the ID column, and save a cleaned copy to
`data/processed/`.